op-program/host/common/l2_store.go

```diff
@@ -0,0 +1,120 @@
+package common
+
+import (
+	"bytes"
+
+	"github.com/ethereum-optimism/optimism/op-program/client/l2"
+	"github.com/ethereum-optimism/optimism/op-program/host/kvstore"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/rawdb"
+	"github.com/ethereum/go-ethereum/ethdb"
+)
+
+type l2KeyValueStore struct {
+	kv kvstore.KV
+}
+
+var _ l2.KeyValueStore = (*l2KeyValueStore)(nil)
+
+// NewL2KeyValueStore creates a l2.KeyValueStore compatible database that's backed by a [kvstore.KV]
+func NewL2KeyValueStore(kv kvstore.KV) *l2KeyValueStore {
+	return &l2KeyValueStore{kv: kv}
+}
+
+var codePrefixedKeyLength = common.HashLength + len(rawdb.CodePrefix)
+
+func unwrapKey(key []byte) []byte {
+	if len(key) == codePrefixedKeyLength && bytes.HasPrefix(key, rawdb.CodePrefix) {
+		return key[len(rawdb.CodePrefix):]
+	}
+	return key
+}
+
+func (db *l2KeyValueStore) Get(key []byte) ([]byte, error) {
+	key = unwrapKey(key)
+	if len(key) != common.HashLength {
+		return nil, l2.ErrInvalidKeyLength
+	}
+	return db.kv.Get(common.Hash(key))
+}
+
+func (db *l2KeyValueStore) Has(key []byte) (bool, error) {
+	key = unwrapKey(key)
+	if len(key) != common.HashLength {
+		return false, l2.ErrInvalidKeyLength
+	}
+	_, err := db.kv.Get(common.Hash(key))
+	switch err {
+	case kvstore.ErrNotFound:
+		return false, nil
+	case nil:
+		return true, nil
+	default:
+		return false, err
+	}
+}
+
+func (db *l2KeyValueStore) Put(key []byte, value []byte) error {
+	key = unwrapKey(key)
+	return db.kv.Put(common.Hash(key), value)
+}
+
+func (db *l2KeyValueStore) NewBatch() ethdb.Batch {
+	return &batch{db: db}
+}
+
+func (db *l2KeyValueStore) NewBatchWithSize(size int) ethdb.Batch {
+	return &batch{db: db}
+}
+
+// batch is similar to memorydb.batch, but adapted for kvstore.KV
+type batch struct {
+	db     *l2KeyValueStore
+	writes []keyvalue
+	size   int
+}
+
+var _ ethdb.Batch = (*batch)(nil)
+
+type keyvalue struct {
+	key   string
+	value []byte
+}
+
+func (b *batch) Put(key []byte, value []byte) error {
+	b.writes = append(b.writes, keyvalue{string(key), common.CopyBytes(value)})
+	b.size += len(key) + len(value)
+	return nil
+}
+
+func (b *batch) Delete(key []byte) error {
+	// ignore deletes
```

Comment:
this should be safe as I don't think it matters for the state transition whether a node is deleted or not. The oracle db [doesn't support deletion](https://github.com/ethereum-optimism/optimism/blob/1cf221b892e2bc30aa3cf3f612d55b682b741617/op-program/client/l2/db.go#L82) either.